File: Classes/Utils/Log.m
/*
 * Copyright (c) 2010-2020 Belledonne Communications SARL.
 *
 * This file is part of linphone-iphone
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

#import "Log.h"
#import <asl.h>
#import <os/log.h>

#ifdef USE_CRASHLYTICS
@import FirebaseCrashlytics;
#endif

@implementation Log

#define FILE_SIZE 17
#define DOMAIN_SIZE 3


+ (void)log:(OrtpLogLevel)severity file:(const char *)file line:(int)line format:(NSString *)format, ... {
	va_list args;
	va_start(args, format);
	NSString *str = [[NSString alloc] initWithFormat:format arguments:args];
	const char *utf8str = [str cStringUsingEncoding:NSString.defaultCStringEncoding];
	const char *filename = strchr(file, '/') ? strrchr(file, '/') + 1 : file;
	ortp_log(severity, "(%*s:%-4d) %s", FILE_SIZE, filename + MAX((int)strlen(filename) - FILE_SIZE, 0), line, utf8str);
	va_end(args);
}

+ (void)enableLogs:(OrtpLogLevel)level {
	BOOL enabled = (level >= ORTP_DEBUG && level < ORTP_ERROR);
	static BOOL stderrInUse = NO;
	if (!stderrInUse) {
		asl_add_log_file(NULL, STDERR_FILENO);
		stderrInUse = YES;
	}
	linphone_core_set_log_collection_path([LinphoneManager cacheDirectory].UTF8String);
	linphone_core_set_log_handler(linphone_iphone_log_handler);
	linphone_core_enable_log_collection(enabled);
	if (level == 0) {
		linphone_core_set_log_level(ORTP_FATAL);
		ortp_set_log_level("ios", ORTP_FATAL);
		NSLog(@"I/%s/Disabling all logs", ORTP_LOG_DOMAIN);
	} else {
		NSLog(@"I/%s/Enabling %s logs", ORTP_LOG_DOMAIN, (enabled ? "all" : "application only"));
		linphone_core_set_log_level(level);
		ortp_set_log_level("ios", level == ORTP_DEBUG ? ORTP_DEBUG : ORTP_MESSAGE);
	}
}

+ (void)directLog:(OrtpLogLevel)level text:(NSString *)text {
	bctbx_log(BCTBX_LOG_DOMAIN, level, "%s", [text cStringUsingEncoding:NSUTF8StringEncoding]);
}

#pragma mark - Logs Functions callbacks

void linphone_iphone_log_handler(const char *domain, OrtpLogLevel lev, const char *fmt, va_list args) {
	NSString *format = [[NSString alloc] initWithUTF8String:fmt];
	NSString *formatedString = [[NSString alloc] initWithFormat:format arguments:args];
	NSString *lvl;

	if (!domain)
		domain = "lib";
	// since \r are interpreted like \n, avoid double new lines when logging network packets (belle-sip)
	// output format is like: I/ios/some logs. We truncate domain to **exactly** DOMAIN_SIZE characters to have
	// fixed-length aligned logs
	switch (lev) {
		case ORTP_FATAL:
			lvl = @"Fatal";
			break;
		case ORTP_ERROR:
			lvl = @"Error";
			break;
		case ORTP_WARNING:
			lvl = @"Warning";
			break;
		case ORTP_MESSAGE:
			lvl = @"Message";
			break;
		case ORTP_DEBUG:
			lvl = @"Debug";
			break;
		case ORTP_TRACE:
			lvl = @"Trace";
			break;
		case ORTP_LOGLEV_END:
			return;
	}
	if ([formatedString containsString:@"\n"]) {
		NSArray *myWords = [[formatedString stringByReplacingOccurrencesOfString:@"\r\n" withString:@"\n"]
			componentsSeparatedByString:@"\n"];
		for (int i = 0; i < myWords.count; i++) {
			NSString *tab = i > 0 ? @"\t" : @"";
			if (((NSString *)myWords[i]).length > 0) {
#ifdef USE_CRASHLYTICS
			[[FIRCrashlytics crashlytics] logWithFormat:@"[%@] %@%@", lvl, tab, (NSString *)myWords[i]];
#endif
			NSLog(@"[%@] %@%@", lvl, tab, (NSString *)myWords[i]);
			}
		}
	} else {
#ifdef USE_CRASHLYTICS
		[[FIRCrashlytics crashlytics] logWithFormat:@"[%@] %@", lvl, [formatedString stringByReplacingOccurrencesOfString:@"\r\n" withString:@"\n"]];
#endif
		NSLog(@"[%@] %@", lvl, [formatedString stringByReplacingOccurrencesOfString:@"\r\n" withString:@"\n"]);
	}
}

@end

File: Linphone/Utils/Log.swift
/*
* Copyright (c) 2010-2023 Belledonne Communications SARL.
*
* This file is part of linphone
*
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program. If not, see <http://www.gnu.org/licenses/>.
*/
// swiftlint:disable line_length

// Singleton instance that logs both info from the App and from the core, using the core log level. ([app] log_level parameter in linphonerc-factory-app

import UIKit
import os
import linphonesw
import linphone
#if USE_CRASHLYTICS
import Firebase
#endif

class Log: LoggingServiceDelegate {
	
	static let instance = Log()
		
	var debugEnabled = true // Todo : bind to app parameters
	var service = LoggingService.Instance

	private init() {
		service.domain = Bundle.main.bundleIdentifier!
		Core.setLogCollectionPath(path: Factory.Instance.getDownloadDir(context: UnsafeMutablePointer<Int8>(mutating: (Config.appGroupName as NSString).utf8String)))
		Core.enableLogCollection(state: LogCollectionState.Enabled)
		setMask()
		LoggingService.Instance.addDelegate(delegate: self)
		
	}
	
	func setMask() {
		if debugEnabled {
			LoggingService.Instance.logLevelMask = UInt(LogLevel.Fatal.rawValue + LogLevel.Error.rawValue + LogLevel.Warning.rawValue + LogLevel.Message.rawValue + LogLevel.Trace.rawValue + LogLevel.Debug.rawValue)
		} else {
			LoggingService.Instance.logLevelMask = UInt(LogLevel.Fatal.rawValue +  LogLevel.Error.rawValue +  LogLevel.Warning.rawValue)
		}
	}
	
	let levelToStrings: [Int: String] =
		[LogLevel.Debug.rawValue: "Debug"
		 , LogLevel.Trace.rawValue: "Trace"
		 , LogLevel.Message.rawValue: "Message"
		 , LogLevel.Warning.rawValue: "Warning"
		 , LogLevel.Error.rawValue: "Error"
		 , LogLevel.Fatal.rawValue: "Fatal"]
	
	let levelToOSleLogLevel: [Int: OSLogType] =
		[LogLevel.Debug.rawValue: .debug,
		 LogLevel.Trace.rawValue: .info,
		 LogLevel.Message.rawValue: .info,
		 LogLevel.Warning.rawValue: .error,
		 LogLevel.Error.rawValue: .error,
		 LogLevel.Fatal.rawValue: .fault]
	
	public class func debug(_ message: String) {
		instance.service.debug(message: message)
	}
	public class func info(_ message: String) {
		instance.service.message(message: message)
	}
	public class func warn(_ message: String) {
		instance.service.warning(message: message)
	}
	public class func error(_ message: String) {
		instance.service.error(message: message)
	}
	public class func fatal(_ message: String) {
		instance.service.fatal(message: message)
	}
	
	private func output(_ message: String, _ level: Int, _ domain: String = Bundle.main.bundleIdentifier!) {
		let log = "[\(domain)][\(levelToStrings[level] ?? "Unkown")] \(message)\n"
		if #available(iOS 10.0, *) {
			os_log("%{public}@", type: levelToOSleLogLevel[level] ?? .info, log)
		} else {
			NSLog(log)
		}
#if USE_CRASHLYTICS
		if FirebaseApp.app() != nil {
			Crashlytics.crashlytics().log(log)
		}
#endif
	}
		
	func onLogMessageWritten(logService: LoggingService, domain: String, level: LogLevel, message: String) {
		output(message, level.rawValue, domain)
	}

	public class func stackTrace() {
		Thread.callStackSymbols.forEach { print($0) }
	}
	
	// Debug
	public class func cdlog(_ message: String) {
		info("cdes>\(message)")
	}
	public class func bmlog(_ message: String) {
		info("bmar>\(message)")
	}
	public class func qelog(_ message: String) {
		info("qarg>\(message)")
	}
		
}

// swiftlint:enable line_length
